Initial version of tower_schedule

john-westcott-iv opened the pull request:
##### SUMMARY
Connect https://github.com/ansible/awx/issues/5461

<!--- Describe the change, including rationale and design decisions -->
Adding tower_schedule build request

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - Collections

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 11.0.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

john-westcott-iv committed:
Initial version of tower_schedule

john-westcott-iv committed:
Adding rrule lookup plugin

beeankha committed:
Make minor changes to fix sanity test results

john-westcott-iv committed:
Adding requirements on pytz and python.dateutil for rrule lookup plugin

beeankha committed:
Fix linter error

beeankha commented:
Connecting Issue https://github.com/ansible/awx/issues/5461 to this PR.

AlanCoding commented on awx_collection/tests/integration/targets/tower_schedule_rrule/tasks/main.yml:
```diff
@@ -0,0 +1,252 @@
+
+- assert:
+    that:
+      - result is not failed
```
these are all fine assertions to keep around for rrule testing, but these are just really high-level tests, and not a good place to put them.

As for the _best_ place for them, it seems like this would be ideal:

https://github.com/ansible/awx/blob/29b567d6e16b620b0eedce1ff8ee2c53828758dd/awx/main/tests/functional/api/test_schedules.py#L128

We are rapid-fire firing off mock requests and asserting that we get the right validation error. These integration tests for the collection do the same thing but in a very verbose way. And honestly these should be unit, not functional tests too, because the database should not be needed either.

So I would recommend that we add 1 test here that hits a 400 error, assert that the right error message is in `result.msg`. Then that should cover all of these cases, as far as the collection is concerned. The specific message is different, but the mechanics on the side of the collection are the same for all.

AlanCoding replied:
oh wait, I wasn't fully understanding, this is a fully independent implementation of rrule logic...

AlanCoding commented on awx_collection/tests/integration/targets/tower_schedule_rrule/tasks/main.yml:
```diff
@@ -0,0 +1,252 @@
+- name: Test too many params
+  debug:
+    msg: "{{ query('awx.awx.tower_schedule_rrule', 'none', 'weekly', start_date='2020-4-16 03:45:07') }}"
+  ignore_errors: true
```
looks like these are all error cases, I would like to see something that tests that it produces something usable by the API. I don't see any argument against combining some happy-path values with the `tower_schedule` module, like what we do in the examples. I know the targets are named so they are separated by modules, but other tests already break that pattern (using other modules in tests for some module), and I say you are free to do so as well.

john-westcott-iv replied:
Other than firing off a bunch of random output combinations, any ideas on how to do positive testing? I did run into some things where the rrule library was not quite "up to snuff" from a Tower perspective. i.e. the lib does not support TZ so I had to do a manual regex replace to get the TZ in there. Also, Tower demands a INTERVAL in the rrule but the library does not add INTERVAL=1 by default so I had to do a concat if interval == 1.

AlanCoding replied:
Do you mean you had to add it here in the lookup plugin itself?

https://github.com/ansible/awx/pull/6733/files#diff-5a73bbe4e04087fee21efbcafe352d8eR230

Ideally the library would do most of this, but I'm worried that you may be getting into a hard problem unexpectedly. @ryanpetrello is the whizz on this.

john-westcott-iv replied:
Yea, that and two more lines down. I actually say a bug report from @ryanpetrello against the rrule lib around this when I was looking for answers. Completely open to suggestions (perhaps there is a different library I should be working with?)

fherbert commented:
I don't see a way of using the rrule lookup plugin to generate a rrule string to run every day at 6:00am. There is no option available to set byhour, byminute and bysecond when settings _terms to day. Any chance this can be added? I'm happy to contribute to the code if it helps.
An example of the rrule string I'd like to be able to generate is:

`DTSTART;TZID=Pacific/Auckland:20190325T000100 RRULE:FREQ=DAILY;INTERVAL=1;BYHOUR=6;BYMINUTE=00`

john-westcott-iv reviewed (commented):
> I don't see a way of using the rrule lookup plugin to generate a rrule string to run every day at 6:00am. There is no option available to set byhour, byminute and bysecond when settings _terms to day. Any chance this can be added? I'm happy to contribute to the code if it helps.
> An example of the rrule string I'd like to be able to generate is:
> 
> `DTSTART;TZID=Pacific/Auckland:20190325T000100 RRULE:FREQ=DAILY;INTERVAL=1;BYHOUR=6;BYMINUTE=00`

@fherbert Thanks for playing around with this new code and giving us feedback. This plugin is intended to function similar to how the Tower UI work. In the UI, if you select daily you get the following fields: `start date, start time, time zone, every and end`. 

Based on that you can create a daily 6am schedule with this: `query('awx.awx.tower_schedule_rrule', 'day', start_date='2019-03-25 06:00:00', timezone='Pacific/Auckland')`

The resulting rule will be: `DTSTART;TZID=Pacific/Auckland:20190325T060000 RRULE:FREQ=DAILY;INTERVAL=1` which is not quite what you were looking to generate but should tell Tower to schedule your job for 6am daily.

Please try that and let us know how it goes.

fherbert commented:
> @fherbert Thanks for playing around with this new code and giving us feedback. This plugin is intended to function similar to how the Tower UI work. In the UI, if you select daily you get the following fields: `start date, start time, time zone, every and end`.
> 
> Based on that you can create a daily 6am schedule with this: `query('awx.awx.tower_schedule_rrule', 'day', start_date='2019-03-25 06:00:00', timezone='Pacific/Auckland')`
> 
> The resulting rule will be: `DTSTART;TZID=Pacific/Auckland:20190325T060000 RRULE:FREQ=DAILY;INTERVAL=1` which is not quite what you were looking to generate but should tell Tower to schedule your job for 6am daily.
> 
> Please try that and let us know how it goes.

@john-westcott-iv, I've tested as you suggested and it works as expected. Also tested the tower_schedule module and it works well. Thanks.

beeankha commented:
recheck

AlanCoding commented on awx_collection/plugins/lookup/tower_schedule_rrule.py:
```diff
@@ -0,0 +1,235 @@
+    missing_modules.append('pytz')
+
+try:
+    from dateutil.rrule import rrule, MONTHLY, WEEKLY, DAILY, HOURLY, MINUTELY, MO, TU, WE, TH, FR, SA, SU
```
I worry about the complexity of content in the `try` for import exceptions like this. If just _one_ of these items moves within the library, then it will begin failing with the wrong message. Tracebacks aren't great for users to go, but they are light years better than _wrong_ information.

Because of that, I would suggest doing a simplified import of `from dateutil import rrule as rrule`, and then referencing these things like `rrule.MONTHLY` in code. There are other options too. Instead of making these imports here, you can do a _test_ import of `import dateutil` and then import those things after the try-except (again, avoids flagrantly wrong error messages).

john-westcott-iv replied:
Fixed.

AlanCoding commented on awx_collection/plugins/lookup/tower_schedule_rrule.py:
```diff
@@ -0,0 +1,235 @@
+        except ValueError:
+            return datetime.strptime(date_string, '%Y-%m-%d')
+
+    def run(self, terms, variables=None, **kwargs):
```
so is this `variables` an artifact of the general plugin interface for lookup plugins? And it's taken here just to not be used?

john-westcott-iv replied:
Correct, that is just how I understand run needs to be declared.

AlanCoding commented on awx_collection/plugins/lookup/tower_schedule_rrule.py:
```diff
@@ -0,0 +1,235 @@
+          - Second string is one of sunday, monday, tuesday, wednesday, thursday, friday
+          - Used for month type schedules
+          - Can not be used with month_day_number parameters
+    extends_documentation_fragment: awx.awx.auth
```
why does it extend this doc fragment? I don't think it should.

john-westcott-iv replied:
Good catch, removed.

john-westcott-iv committed:
Changing import

john-westcott-iv committed:
Remocing doc fragment extension

AlanCoding committed:
Unit testing of tower_schedule

Move previously integration tests of lookup plugin to unit tests
  delete all integration tests except some basic demo tests

do simple create unit test

AlanCoding commented:
Heads up, I put up https://github.com/john-westcott-iv/awx/pull/6/files, should be rebased right now

john-westcott-iv committed:
Adding validation of rrule output thorugh awx searlizer

AlanCoding commented:
recheck

john-westcott-iv commented:
recheck

beeankha committed:
Update Makefile to find correct python path for testing

beeankha commented:
recheck

beeankha commented:
recheck

AlanCoding reviewed (approved):
This is really good quality work, thanks!

squidboylan commented:
Looks like we're close to getting this in but i have 2 concerns.

It looks like we do a lot of `ignore_errors: True` in the tests when it's not necessary, which could lead to false positives, and if we expect the test to pass we should fail if it fails rather than ignoring errors then doing an assert, this provides better error messages rather than "Assertions failed'.

The other thing is the `tower_schedule_rrule` plugin tests fail on rhel 7 and rhel 8 because the system dateutil is old and leaves out the `RRULE:` bit. We should either detect the old dateutil and have an error message that says that's why it's breaking (debugging this otherwise is nontrivial), or make it work with the system dateutil which is 2.6.1 on rhel 8 and 1.5 on rhel 7.

squidboylan commented on awx_collection/tests/integration/targets/tower_schedule/tasks/main.yml:
```diff
@@ -0,0 +1,91 @@
+    unified_job_template: "Demo Job Template"
+    rrule: "DTSTART:20191219T130551Z RRULE:FREQ=WEEKLY;INTERVAL=1;COUNT=1"
+  register: result
+  ignore_errors: true
```
i dont think we need this

squidboylan commented on awx_collection/tests/integration/targets/tower_schedule/tasks/main.yml:
```diff
@@ -0,0 +1,91 @@
+    unified_job_template: "Demo Job Template"
+    rrule: "DTSTART:20191219T130551Z RRULE:FREQ=WEEKLY;INTERVAL=1;COUNT=1"
+  register: result
+  ignore_errors: true
```
i dont think we need this

squidboylan commented on awx_collection/tests/integration/targets/tower_schedule/tasks/main.yml:
```diff
@@ -0,0 +1,91 @@
+    state: present
+    enabled: "false"
+  register: result
+  ignore_errors: true
```
i dont think we need this

squidboylan reviewed (changes requested):
left some inline comments about `ignore_errors: true`

john-westcott-iv committed:
Removing ignore_errors for tasks which should succeed

john-westcott-iv committed:
Validating the version of python-dateutil

john-westcott-iv commented:
@squidboylan ignore_errors and dateutil version should be all set with these two new commits.

squidboylan reviewed (changes requested):
I just tested locally and it looks like we need `python.dateutil >= 2.7.0`. I also cant tell how that try block works, looks like it will only fail if dateutil doesnt exist, I think we want to fail early if we have an incompatible version.

AlanCoding commented:
Do you really want to _fail_ with a version too old, or just give a warning? Also, `dateutil.__version__` may not necessarily work correctly with all versions, so something like `getattr(dateutil, '__version__', 'unknown')` may be better.

I don't expect that _all_ uses of this plugin will fail with older versions, and I don't think it's standard Ansible practice to fail when using libs with older versions.

I was going to suggest:

```diff
diff --git a/awx_collection/plugins/lookup/tower_schedule_rrule.py b/awx_collection/plugins/lookup/tower_schedule_rrule.py
index 96bd1735e8..6d19a807da 100644
--- a/awx_collection/plugins/lookup/tower_schedule_rrule.py
+++ b/awx_collection/plugins/lookup/tower_schedule_rrule.py
@@ -231,8 +231,12 @@ class LookupModule(LookupBase):
                 raise AnsibleError('Timezone parameter is not valid')
             timezone = kwargs['timezone']
 
-        # rrule puts a \n in the rule instad of a space and can't hand timezones
-        return_rrule = str(my_rule).replace('\n', ' ').replace('DTSTART:', 'DTSTART;TZID={0}:'.format(timezone))
+        # rrule puts a \n in the rule instad of a space and can't handle timezones
+        rrule_string = str(my_rule)
+        separator_string = ' '
+        if 'RRULE:' not in rrule_string:
+            separator_string = ' RRULE:'
+        return_rrule = rrule_string.replace('\n', separator_string).replace('DTSTART:', 'DTSTART;TZID={0}:'.format(timezone))
         # Tower requires an interval. rrule will not add interval if its set to 1
         if kwargs.get('every', 1) == 1:
             return_rrule = "{0};INTERVAL=1".format(return_rrule)
```

This may help _expand_ the working version envelope, but only by a limited amount, and I wouldn't suggest changing the version suggestion as it is now.

squidboylan commented:
> Do you really want to fail with a version too old, or just give a warning?

> I don't expect that all uses of this plugin will fail with older versions.

I was thinking most usescases for this would fail with older dateutil but i would be happy with at least a warning.

> This may help expand the working version envelope, but only by a limited amount, and I wouldn't suggest changing the version suggestion as it is now.

Yeah we could include this but probably still want a dateutil>=2.7.0 recommendation

john-westcott-iv committed:
Now actually checking the version instead of just getting it

AlanCoding commented on awx_collection/plugins/modules/tower_schedule.py:
```diff
@@ -0,0 +1,252 @@
+    unified_job_template: "Demo Job Template"
+    rrule: "DTSTART:20191219T130551Z RRULE:FREQ=WEEKLY;INTERVAL=1;COUNT=1"
+  register: result
+  ignore_errors: True
```
why do these examples have `ignore_errors`? Deleting that might be best for clarity

john-westcott-iv replied:
Good catch, I think I was playing around with extracting the example docs out of our unit testing and those were in there. Should be removed from both now.

AlanCoding commented:
Looking at update:

 - versioning is handled coherently, 👍 for current code from me
 - looks like `ignore_errors` in tests was updated

left non-blocking comment about example.

Looks ready for re-testing to me.

john-westcott-iv committed:
Removing ignore_errors from examples

squidboylan reviewed (approved):
:shipit:
